Wikipedia/UI-V5/WMFArticleHeaderImageGalleryViewController.m

```diff
@@ -95,45 +98,141 @@ - (UICollectionViewCell*)collectionView:(UICollectionView*)collectionView
         cell.imageView.contentMode = UIViewContentModeScaleAspectFill;
         return cell;
     }
+
     MWKImage* imageMetadata = self.images[indexPath.item];
+
+    // try to set cell's image from memory cache, otherwise fetch it and do face detection if needed
+    if (![self setCachedImageForCell:cell atIndexPath:indexPath metadata:imageMetadata]) {
+        [self fetchImage:imageMetadata forCellAtIndexPath:indexPath];
```

Comment:
Perhaps, but then what do you do if fact detection has run, but the image isn't in memory? Simpler to have one idempotent code path. This is one of the reasons I tried to avoid a separate fast path for in-memory, face-detected images, but it was shown that the optimization was worth it, so here we are :smile:.